File: WebKit/Examples/ajaxpage.js
/*
	Basic Ajax Javascript functions used by AjaxPage.

	Written by John Dickinson based on ideas from
	Apple developer code (developer.apple.com)
	and Nevow 0.4.1 (www.nevow.org).
*/

var request_count = 0;

function createRequest()
{
	var req;
	if (window.XMLHttpRequest) {
		req = new XMLHttpRequest();
	} else if (window.ActiveXObject) {
		req = new ActiveXObject("Microsoft.XMLHTTP");
	}
	return req;
}

function openConnection(req, url)
{
	if (req) {
		req.onreadystatechange = function() {
			if (req.readyState == 4) {
				if (req.status == 200) {
					try {
						eval(req.responseText);
					} catch (e) {
						; // ignore errors
					}
				}
			}
		};
		req.open("GET", url, true);
		req.send(null);
	}
}

// Generic Ajax call:
function generic_ajax(pre_action, func) {
	if (pre_action) {
		eval(pre_action);
	}
	var additionalArguments = ''
	for (i = 2; i<arguments.length; i++) {
		additionalArguments += '&a='
		additionalArguments += encodeURIComponent(arguments[i])
	}
	request_count += 1;
	additionalArguments += '&a=' + request_count;
	var loc = document.location.toString();
	if (loc.indexOf('?') != -1) {
		loc = loc.substr(0, loc.indexOf('?'));
	}
	req = createRequest();
	if (req) {
		openConnection(req, loc+"?_action_=ajax_controller&f="+func+additionalArguments);
	}
}

// Ajax call specific to forms:
function generic_ajax_form(func, f, dest) {
	if (dest) {
		var hf = document.getElementById(dest);
		hf.innerHTML = '<img src="images/animateddots.gif" alt="Your form is being processed. Please be patient."/>';
	}
	var values = Array();
	for (i = 0; i<f.elements.length; i++) {
		var e = f.elements[i];
		name = e.name;
		if (!(((e.type == 'checkbox') || (e.type == 'radio')) && (!e.checked))) {
			values[i] = [name, e.value];
		}
	}
	var additionalArguments = ''
	for (i = 0; i<values.length; i++) {
		additionalArguments += '&a='
		additionalArguments += encodeURIComponent(values[i])
	}
	request_count += 1;
	additionalArguments += '&a=' + request_count;
	var loc = document.location.toString();
	if (loc.indexOf('?') != -1) {
		loc = loc.substr(0, loc.indexOf('?'));
	}
	req = createRequest();
	if (req) {
		openConnection(req, loc+"?_action_=ajax_controller&f="+func+additionalArguments);
	}
}

// Below are some Ajax helper functions:

function ajax_setTag(which, val) {
	var e = document.getElementById(which);
	e.innerHTML = val;
}

function ajax_setClass(which, val) {
	var e = document.getElementById(which);
	e.className = val;
}

function ajax_setID(which, val) {
	var e = document.getElementById(which);
	e.id = val;
}

function ajax_setValue(which, val) {
	var e = document.getElementById(which);
	e.value = val;
}

function ajax_setReadonly(which, val) {
	var e = document.getElementById(which);
	if (val) {
		e.setAttribute('readonly', 'readonly');
	}
	else {
		e.removeAttribute('readonly');
	}
}

File: WebKit/Examples/AjaxPage.py
#
# AjaxPage.py
#
# Written by John Dickinson based on ideas from
# Apple developer code (developer.apple.com)
# and Nevow 0.4.1 (www.nevow.org).
# Minor changes by Robert Forkel and Christoph Zwerschke.
#

import StringIO, traceback, time, random

from ExamplePage import ExamplePage as BaseClass

# PyJavascript and quote_js based on ideas from Nevow 0.4.1 (www.nevow.org)

def quote_js(what):
	"""Return quoted JavaScript string corresponding to the Python object."""
	if isinstance(what, bool):
		ret = str(what).lower()
	elif isinstance(what, (int, long, float, PyJavascript)):
		ret = str(what)
	else:
		ret = "'%s'" % str(what).replace('\\', '\\\\').replace('\'', '\\\'').replace('\n', '\\n')
	return ret


class PyJavascript(object):
	"""This class simply tanslates a Python expression into a JavaScript string."""

	def __init__(self, name):
		self.__name = name

	def __getattr__(self, aname):
		return self.__class__('%s.%s'%(self, aname))

	def __str__(self):
		return self.__name

	def __call__(self, *a, **kw):
		args = ','.join([quote_js(i) for i in a])
		kwargs = ','.join(['%s=%s' % (k, quote_js(v)) for k, v in kw.items()])
		if args and kwargs:
			allargs = '%s,%s' % (args, kwargs)
		elif not kwargs:
			allargs = args
		elif not args:
			allargs = kwargs
		return self.__class__('%s(%s)' % (self, allargs))

	def __getitem__(self, index):
		return self.__class__('%s[%s]' % (self, quote_js(index)))

	def __repr__(self):
		return self.__str__()


class AjaxPage(BaseClass):
	"""A superclass for Webware servlets using Ajax techniques.

	AjaxPage can be used to make coding XMLHttpRequest() applications easier.

	Subclasses should override the method ajax_allowed() which returns a list
	of method names. These method names refer to Webware Servlet methods that
	are able to be called by an Ajax-enabled web page. This is very similar
	in functionality to Webware's actions.

	A polling mechanism can be used for long running requests (e.g. generating
	reports) or if you want to send commands to the client without the client
	first triggering an event (e.g. for a chat application). In the first case,
	you should also specify a timeout after which polling shall be used.

	"""

	# Class level variables that can be overridden by servlet instances:
	debug = 0 # set to True if you want to see debugging output
	clientPolling = 1 # set to True if you want to use the polling mechanism
	responseTimeout = 90 # timeout of client waiting for a response in seconds

	# Class level variables to help make client code simpler:
	document = PyJavascript('document')
	setTag = PyJavascript('ajax_setTag')
	setClass = PyJavascript('ajax_setClass')
	setValue = PyJavascript('ajax_setValue')
	setReadonly = PyJavascript('ajax_setReadonly')
	alert = PyJavascript('alert')
	generic_ajax = PyJavascript('generic_ajax')
	generic_ajax_form = PyJavascript('generic_ajax_form')
	this = PyJavascript('this')

	# Response Queue for timed out queries:
	_responseQueue = {}

	def writeJavaScript(self):
		BaseClass.writeJavaScript(self)
		s = '<script type="text/javascript" src="ajaxpage%s.js"></script>'
		self.writeln(s % '')
		if self.clientPolling:
			self.writeln(s % '2')

	def actions(self):
		a = BaseClass.actions(self)
		a.append('ajax_controller')
		if self.clientPolling:
			a.append('ajax_response')
		return a

	def ajax_allowed(self):
		return []

	def ajax_clientPollingInterval(self):
		"""Set the interval for the client polling.

		You should always make it a little random to avoid synchronization.

		"""
		return random.choice(range(3, 8))

	def ajax_response(self):
		"""Return queued Javascript functions to be executed by the client.

		This is polled by the client in random intervals in order to get
		results from long-running queries or push content to the client.

		"""
		if self.clientPolling:
			who = self.session().identifier()
			# Set the timeout until the next time this function is called
			# by the client, using the Javascript wait variable:
			cmd = 'wait=%s;' % self.ajax_clientPollingInterval()
			if self._responseQueue.get(who, []): # add in other commands
				cmd += ';'.join([str(val) for req_number, val in self._responseQueue[who]])
				self._responseQueue[who] = []
			if self.debug:
				self.log("Ajax returns from queue: " + cmd)
			self.write(cmd) # write out at least the wait variable

	def ajax_controller(self):
		"""Execute function f with arguments a on the server side.

		Returns Javascript function to be executed by the client immediately.

		"""
		fields = self.request().fields()
		func = fields.get('f')
		args = fields.get('a', [])
		if type(args) != type([]):
			args = [args]
		req_number = args.pop()
		if self.clientPolling and self.responseTimeout:
			start_time = time.time()
		val = self.alert('There was some problem!')
		if func in self.ajax_allowed():
			try:
				func_obj = getattr(self, func)
			except AttributeError:
				val = self.alert('%s, although an approved function, was not found' % func)
			else:
				try: # pull off sequence number added to "fix" IE
					if self.debug:
						self.log("Ajax call %s(%s)" % (func, args))
					val = str(func_obj(*args))
				except Exception:
					err = StringIO.StringIO()
					traceback.print_exc(file=err)
					e = err.getvalue()
					val = self.alert('%s was called, but encountered an error: %s'%(func,e))
					err.close()
		else:
			val = self.alert('%s is not an approved function' % func)
		if self.clientPolling and self.responseTimeout:
			in_time = time.time() - start_time < self.responseTimeout
		else:
			in_time = 1
		if in_time:
			# If the computation of the function did not last very long,
			# deliver it immediately back to the client with this response:
			if self.debug:
				self.log("Ajax returns immediately: " + str(val))
			self.write(val)
		else:
			# If the client request might have already timed out,
			# put the result in the queue and let client poll it:
			if self.debug:
				self.log("Ajax puts in queue: " + str(val))
			who = self.session().identifier()
			if not self._responseQueue.has_key(who):
				self._responseQueue[who] = [(req_number, val)]
			else:
				self._responseQueue[who].append((req_number, val))

	def ajax_cmdToClient(self, cmd):
		"""Send Javascript commands to the client.

		Client polling must be activitated if you want to use this.

		"""
		if self.clientPolling:
			who = self.session().identifier()
			if not self._responseQueue.has_key(who):
				self._responseQueue[who] = []
			self._responseQueue[who].append((None, cmd))

	def preAction(self, action_name):
		if action_name.startswith('ajax_'):
			pass
		else:
			BaseClass.preAction(self, action_name)

	def postAction(self, action_name):
		if action_name.startswith('ajax_'):
			pass
		else:
			BaseClass.postAction(self,action_name)
